### PR TITLE
Failed authentication are now properly logged

### PR DESCRIPTION
**Pull request by daviian**
Targets #2301 

Failed authentication now returns 401 status code.
Done by introducing a new RenderFormWithErr function in context to allow passing of a statuscode

Tested in following browsers:
- Latest version of FF, Chrome and Safari Mac
- IE 11
- Chrome for Android Latest
- Microsoft Edge 12

### CODE REVIEW

**Comment by lafriks** [comment]
LGTM

**Comment by lunny** [comment]
It seems CI status is disappeared?

**Comment by lafriks** [comment]
@lunny yes it was there until recent force push when he fixed my suggested in discord changes :) Don't know how Drone status got lost

File: integrations/signin_test.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+		errorMessage string
+	}{
+		{username: "wrongUsername", password: "password", errorMessage: i18n.Tr("en", "form.username_password_incorrect")},
+		{username: "user15", password: "wrongPassword", errorMessage: i18n.Tr("en", "form.username_password_incorrect")},
+		{username: "wrongUsername", password: "wrongPassword", errorMessage: i18n.Tr("en", "form.username_password_incorrect")},
+		{username: "user15@example.com", password: "password", errorMessage: i18n.Tr("en", "form.email_been_used")},
+		{username: "user1@example.com", password: "wrongPassword", errorMessage: i18n.Tr("en", "form.username_password_incorrect")},
+	}
+
+	prepareTestEnv(t)
```
Comment (daviian) [inline integrations/signin_test.go]: Shouldn't this be in the testLoginFailed function because each call represents a seperate test case!?

Comment (lafriks) [reply integrations/signin_test.go]: if test-case would change data that need to be reset than yes, otherwise it will only make tests run slower.

File: models/fixtures/user.yml [hunk 2]
```diff
@@ -218,3 +218,33 @@
   avatar_email: user13@example.com
   num_repos: 3
   is_active: true
+
```
Comment (ethantkoenig) [inline models/fixtures/user.yml]: Is there a reason we can't just use existing user fixtures? user16 isn't used anywhere, and it seems we could use any existing user in place of user15 in `TestSignin`.

File: modules/context/context.go [hunk 3]
```diff
@@ -81,14 +81,18 @@ func (ctx *Context) HTML(status int, name base.TplName) {
...
+// RenderFormWithErr used for pages with failed form validation to render flash message for users.
+func (ctx *Context) RenderFormWithErr(status int, msg string, tpl base.TplName, form interface{}) {
 	if form != nil {
 		auth.AssignForm(form, ctx.Data)
 	}
-	ctx.Flash.ErrorMsg = msg
-	ctx.Data["Flash"] = ctx.Flash
-	ctx.HTML(200, tpl)
+	ctx.Flash.Error(msg, true)
+	ctx.HTML(status, tpl)
```
Comment (ethantkoenig) [inline modules/context/context.go]: Why did you remove the `ctx.Data["Flash"] = ctx.Flash` line?

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
I'm not sure returning a 401 on failed login is the correct thing to do; see https://stackoverflow.com/questions/8273757/should-a-failed-login-attempt-result-in-a-http-401-response

I've investigated several public-facing sites (Google, Amazon, Instagram, Github, Gitlab), and none of them returned 401 on failed login.

File: modules/context/context.go (hunk 3 above)
Comment (lafriks) [reply modules/context/context.go]: @ethantkoenig `ctx.Flash.Error(...)` sets `ctx.Data["Flash"]`

Comment (daviian) [reply modules/context/context.go]: @lafriks @ethantkoenig Only if second parameter is `true`

File: models/fixtures/user.yml (hunk 2 above)
Comment (daviian) [reply models/fixtures/user.yml]: Didn't want to intervene in other test cases because of duplicate email address.
But if that's not the case of course I can remove user16.

**Comment by daviian** [comment]
The main reason for 401 status code was to enable easier fail2ban configuration.
An alternative would be to only log the failed authentication attempt in an easy to distinguish way into the gitea log if that pleases you more ;-)

File: models/fixtures/user.yml (hunk 2 above)
Comment (ethantkoenig) [reply models/fixtures/user.yml]: Oh, I missed that user15 and user16 had the same email addresses; sorry about that.

Instead of introducing two new fixtures for testing the duplicate email case, could we instead do something like:
```
func TestDuplicateEmailFailedLogin(t *testing.T) {
	prepareTestEnv(t)
	user := models.AssertExistsAndLoadBean(t, &models.User{ID: 2}).(*models.User)

	// add new user with user2's email
	user.Name = "testuser"
	user.LowerName = strings.ToLower(user.Name)
	user.ID = 0 // will be auto-populated on insert
	models.AssertSuccessfulInsert(t, user)
	
	// TODO try to login with user.Email, should fail with duplicate email
}
```

It's not really scalable for us to add new test fixtures for every corner case we need to test.

**Comment by lafriks** [comment]
I'm now thinking that entry to gitea log would be better than changing response status code

**Comment by ethantkoenig** [comment]
@daviian @lafriks I would prefer a log message as well

File: models/fixtures/user.yml (hunk 2 above)
Comment (daviian) [reply models/fixtures/user.yml]: If others feel the same way I will change that ;-)

**Comment by daviian** [comment]
@lafriks @ethantkoenig Then it's settled. I will change PR.

**Comment by daviian** [comment]
@lafriks @ethantkoenig Which log level would you say is appropriate? Or is there a way to log independent from the set log level?

**Comment by lafriks** [comment]
@ethantkoenig but than fail2ban can be only used by enabling only at least info log level that is not best option. And in case of fail2ban usage failed authorization is actionable entry

**Comment by daviian** [comment]
@ethantkoenig @lafriks I was also struggling with this, but I think users who want to use fail2ban have to set the necessary log level either way. And I agree that `Info` is the best choice.

**Comment by lafriks** [comment]
Why did you remove integration tests? These were useful anyway

**Comment by daviian** [comment]
I thought that they are not part of the PR if it is only logging. But I can add them again.

**Comment by lafriks** [comment]
You could at least submit them in other PR then, we need tests :)

**Comment by daviian** [comment]
And regarding the log level. I am open for discussion ;-) It's not cast in stone yet ;-)

**Comment by lafriks** [comment]
LGTM

**Comment by lafriks** [comment]
@ethantkoenig need your approval

**Comment by bkcsoft** [comment]
Changed the title to reflect the change 🙂

**Review by ethantkoenig — APPROVED** [review]
LGTM

**Comment by ethantkoenig** [comment]
Let's try again LGTM
